From c2fc11231a50c1b3c5353549509a6543b46d35cb Mon Sep 17 00:00:00 2001
From: chdugw_auto <jkulkarni@maxlinear.com>
Date: Fri, 6 Aug 2021 16:28:52 +0300
Subject: [PATCH] Merge pull request #3440 in SW_UGW/linux from
 feature/PONRTSYS-8997-lgm-pon-multicast-support-master to master

* commit 'c85ed09bd0c3968f24ac4f9c2709f0d5f793294d':
  PONRTSYS-8997:URX PON multicast support
---
 drivers/net/datapath/pon_qos/pon_qos_tc_ip_drop.c | 22 ++++++++++++++++++----
 1 file changed, 18 insertions(+), 4 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_ip_drop.c b/drivers/net/datapath/pon_qos/pon_qos_tc_ip_drop.c
index c160097a2171..7ef8b56719b2 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_ip_drop.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_ip_drop.c
@@ -33,7 +33,7 @@ static int parse_mask(struct net_device *dev,
 
 	/* Convert mask to nimble */
 	for (m = 0; m < mask_size; m++) {
-		mask_cpy = mask[m];
+		mask_cpy = htonl(mask[m]);
 		for (i = 0; i < 4; i++, idx += 2) {
 			if (!(mask_cpy & 0x0f))
 				rule->pattern.nDstIP_Mask |= 1 << idx;
@@ -43,6 +43,13 @@ static int parse_mask(struct net_device *dev,
 		}
 	}
 
+	/* Reorder bytes in case of IPv6 masks because network byte order
+	 * is expected. Not required in case of IPv4 as only the lowest
+	 * significant byte is used and all others are ignored.
+	 */
+	if (mask_size == 4)
+		rule->pattern.nDstIP_Mask = htonl(rule->pattern.nDstIP_Mask);
+
 	netdev_dbg(dev, "%s: nDstIP_Mask: %x\n",
 		   __func__, rule->pattern.nDstIP_Mask);
 
@@ -67,7 +74,7 @@ static int parse_ipv4(struct net_device *dev,
 					 f->mask);
 
 	rule->pattern.eDstIP_Select = GSW_PCE_IP_V4;
-	memcpy(&rule->pattern.nDstIP.nIPv4, &key->dst, sizeof(key->dst));
+	rule->pattern.nDstIP.nIPv4 = htonl(key->dst);
 	if (ipv4_is_multicast(key->dst))
 		rule->action.ePortFilterType_Action =
 			GSW_PCE_PORT_FILTER_ACTION_6;
@@ -81,6 +88,7 @@ static int parse_ipv6(struct net_device *dev,
 {
 	struct flow_dissector_key_ipv6_addrs *key;
 	struct flow_dissector_key_ipv6_addrs *mask;
+	int i;
 
 	if (!dissector_uses_key(f->dissector, FLOW_DISSECTOR_KEY_IPV6_ADDRS))
 		return -EIO;
@@ -93,12 +101,18 @@ static int parse_ipv6(struct net_device *dev,
 					 f->mask);
 
 	rule->pattern.eDstIP_Select = GSW_PCE_IP_V6;
-	memcpy(rule->pattern.nDstIP.nIPv6, &key->dst, sizeof(key->dst));
+	/* IPv6 addresses inside structure GSW_PCE_rule_t are stored
+	 * in an array of 16bit values
+	 */
+	for (i = 0; i < 8; ++i) {
+		((uint16_t *)rule->pattern.nDstIP.nIPv6)[i] =
+			htons(((uint16_t *)&key->dst)[i]);
+	}
 	if (ipv6_addr_is_multicast(&key->dst))
 		rule->action.ePortFilterType_Action =
 			GSW_PCE_PORT_FILTER_ACTION_6;
 
-	return parse_mask(dev, rule, key->dst.in6_u.u6_addr32, 4);
+	return parse_mask(dev, rule, mask->dst.in6_u.u6_addr32, 4);
 }
 
 static int parse_ip(struct net_device *dev,
